File: Week 6 Python/sentimental-hello/credit.py
import math


def get_credit():
    credit = 0
    while True:
        try:
            credit = int(input("Input Credit Card Number: "))
        except ValueError:
            print("Invalid Card Number.")
            continue
        else:
            break
    return credit


def check_digit():
    # credit = get_credit()
    sum = 0
    breakdown = 0
    for i in range(16):
        divisor = pow(10, i)
        digit = credit // divisor
        lastDigit = digit % 10
        multiplier = lastDigit * 2

        if i % 2 == 0:
            sum += lastDigit
        else:
            if multiplier <= 9:
                sum += multiplier
            else:
                breakdown = (multiplier % 10) + (multiplier // 10 % 10)
                sum += breakdown

    return sum


credit = get_credit()
sum = check_digit()
checkDigit = math. ceil(math.log10(credit))
first_digit = credit // pow(10, checkDigit - 1)
first_two_digit = credit // pow(10, checkDigit - 2)

if sum % 10 == 0:
    if checkDigit == 15 and (first_two_digit == 34 or first_two_digit == 37):
        print("AMEX")

    elif checkDigit == 16 and (first_two_digit == 51 or first_two_digit == 52 or first_two_digit == 53 or first_two_digit == 54 or first_two_digit == 55):
        print("MASTERCARD")

    elif first_digit == 4 and (checkDigit == 13 or checkDigit == 16):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: Week 1 C/Credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main()
{
    long cr = get_long("Number: ");
    int checkDigit = (log10(cr) + 1);
    int sum = 0;
    for (int i = 0; i <= 15; i++)
    {
        long divisor = pow(10, (i));
        long digit = cr / divisor;
        long lastDigit = digit % 10;
        long multiplier = lastDigit * 2;

        if (i % 2 == 0)
        {
            sum += lastDigit;
        }
        else
        {
            if (multiplier <= 9)
            {
                sum += multiplier;
            }
            else
            {
                long breakdown = (multiplier % 10) + (multiplier / 10 % 10);
                sum += breakdown;
            }
        }
    }
    long first_digit = cr / pow(10, checkDigit - 1);
    long first_two_digit = cr / pow(10, checkDigit - 2);
    // check card details
    if (sum % 10 == 0)
    {
        if (checkDigit == 15 && (first_two_digit == 34 || first_two_digit == 37))
        {
            printf("AMEX\n");
        }
        else if (checkDigit == 16 &&
                 (first_two_digit == 51 || first_two_digit == 52 || first_two_digit == 53 ||
                  first_two_digit == 54 || first_two_digit == 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((checkDigit == 13 || checkDigit == 16) && first_digit == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
